### ISSUE TITLE
Reconsider both curried and tupled overloads on dsharp.add etc.

### ISSUE THREAD

**Issue by dsyme** [DiffSharp/DiffSharp#155]
This pattern of having both curried and tupled methods is not recommended F# design and will likely cause problems (I noticed it in strange doc generation).  I think we will probably have to just choose one or another

```
    static member add(a:Tensor, b:Tensor) = a.add(b)
    static member add(b:Tensor) = fun (a:Tensor) -> a.add(b)
    static member sub(a:Tensor, b:Tensor) = a.sub(b)
    static member sub(b:Tensor) = fun (a:Tensor) -> a.sub(b)
    static member mul(a:Tensor, b:Tensor) = a.mul(b)
    static member mul(b:Tensor) = fun (a:Tensor) -> a.mul(b)
    static member div(a:Tensor, b:Tensor) = a.div(b)
    static member div(b:Tensor) = fun (a:Tensor) -> a.div(b)
```

**Comment by gbaydin** [comment]
I would really like to keep this if possible.

As I'm writing DiffSharp code, supporting both styles gave me great flexibility. The curried ones work very well with functional F# style. For example here:

https://github.com/DiffSharp/DiffSharp/blob/33dd61eff9eb33460158c4eee98f744390402630/src/DiffSharp.Core/Distributions.fs#L202

But at the same time I am also able to use curried style for more PyTorch-like code. Is the problem only about the documentation?

**Comment by dsyme** [comment]
> Is the problem only about the documentation?

There are multiple potential problems:

1. Order of arguments is confusing, e.g. `dsharp.sub x y` is `y - x` not `x - y`. That's really very confusing to people, and there's nothing to force currying to be used.

2. F# type inference and tooling is not really designed for APIs like this.  I just don't know if it will be that great. Certainly F# type inference just works much better if there are fewer overloads.  Also will parameter help work? I don't know.

3. 2x the documentation

4. It might be fine for div/add/mul/... etc. with two arguments but get whacky with other things that have more arguments. What's the general rule?

5. Intellisense will be more confusing to people - 2x the entries and they won't know which to choose

6. If we add shape inference, then we have 2x the API surface area

7. Too many ways to do the same thing

8. Are user-defined binary operators also meant to follow this pattern?

Of these (1) is to me too severe to embrace this on anything non-commutative.

There's commonly also a problem where things like this combine: a 2x multiplication of your API surface area becomes 4x then 8x.... when combined with something else (like overloading on scalars).

Anyway I'd recommend against building this in for now for all the above reason. In your example just write the lambda

    else d.valuesTensor.unstack() |> Seq.map f |> dsharp.stack |> (fun t -> t.mean(0))

and separately advocate for solving this at the F# language level, e.g. via `_.x(args)` notation:

    else d.valuesTensor.unstack() |> Seq.map f |> dsharp.stack |> _.mean(0)

**Comment by gbaydin** [comment]
Ok in that case I think we should prefer the tupled methods and deprecate the curried versions. I can go ahead and make a PR with this change.

**Comment by gbaydin** [comment]
As a note the design for `sub` was as follows:

```fsharp
dsharp.sub(x, y) // x - y
```

alternatively
```fsharp
x |> dsharp.sub y // x - y
```

**Comment by gbaydin** [comment]
@dsyme some other nice examples were:

https://github.com/DiffSharp/DiffSharp/blob/33dd61eff9eb33460158c4eee98f744390402630/src/DiffSharp.Core/Data.fs#L83
```fsharp
r.ReadBytes(n * rows * cols)
|> Array.map float32
|> dsharp.tensor
|> dsharp.view ([n; 1; 28; 28])
```

https://github.com/DiffSharp/DiffSharp/blob/33dd61eff9eb33460158c4eee98f744390402630/src/DiffSharp.Core/Data.fs#L32
```fsharp
data |> dsharp.stack |> dsharp.move(dtype, device, backend)
```

I was really happy this coding style was available at the same time with the other one.

**Comment by gbaydin** [comment]
More examples I was really happy about
:)

https://github.com/DiffSharp/DiffSharp/blob/33dd61eff9eb33460158c4eee98f744390402630/tests/DiffSharp.Tests/TestModel.fs#L91

https://github.com/DiffSharp/DiffSharp/blob/33dd61eff9eb33460158c4eee98f744390402630/tests/DiffSharp.Tests/TestModel.fs#L284

**Comment by dsyme** [comment]
I do understand the tension here, though I'm not sure it's resolvable.  It's the same problem for any F# API that benefits from currying - either you use currying or tupling, but mixing the two causes really significant problems.

> As a note the design for sub was as follows...

Right, though that assumes the user knows that this is the intended use of the curried function, and chances are they won't (and the bugs are subtle...).  

Unfortunately F#'s `(-)` or `Array.append` for example take their first argument first, and set examples for how currying is going to work with non-commutative binary operators.  If there was a way to restrict `sub` to be used only in pipelining then I get what you're saying but there isn't (and people will really not know this, there's nothing to tell people).

I'm not sure what to suggest.  A compromise may not be too bad. How much of the `dsharp.*` API is affected, and how about simply use currying for the ones you care about (but not `sub` and `div`)?  For example do `dsharp.view` and `dsharp.move` really need to have their tupled variations?  The dot notation normally suffices for them in any case?

I agree this does look nice: 

```
let net = dsharp.view [-1; 10] --> Linear(10, 32) --> dsharp.relu --> Linear(32, 10)
```

**Comment by kevmal** [comment]
> Unfortunately F#'s (-) or Array.append for example take their first argument first, and set examples for how currying is going to work with non-commutative binary operators. If there was a way to restrict sub to be used only in pipelining then I get what you're saying but there isn't (and people will really not know this, there's nothing to tell people).

Non-commutative binary operators pose a general challenge with piping style, since half the time the arg order won't work in your favor. One option would be to include both, for example `dsharp.divBy` and `dsharp.divFrom`, and maybe the naming will prevent assumptions being made on intended use (with `|>` or direct).  Generally I would think the tupled from would not see much use (just use `(-)` or `(/)`, which is much clearer) whereas the curried form may often be a convenient option.

**Comment by dsyme** [comment]
> Non-commutative binary operators pose a general challenge with piping style, since half the time the arg order won't work in your favor. One option would be to include both, for example dsharp.divBy and dsharp.divFrom, and maybe the naming will prevent assumptions being made on intended use (with |> or direct). Generally I would think the tupled from would not see much use (just use (-) or (/), which is much clearer) whereas the curried form may often be a convenient option.

Right - it feels like judicious, careful renaming (with selective use of currying) in `dsharp.*` will be the way to go here.

**Comment by dsyme** [comment]
@gbaydin I do agree that arranging things so that `Tensor -> Tensor` functions are a significant unit of composition in the API is an important consideration.  It's just so natural and simple.

**Comment by dsyme** [comment]
Proposal from discussion with @gbaydin  - make an explicit open to get the pipelined notation


```
    t.dropout()
    t.dropout(0.5)

    dsharp.dropout (t, 0.5) // never used in documentation (curried, no pipeline)

    open DiffSharp.Pipelined
    t |> dsharp.dropout()
    t |> dsharp.dropout(0.5)
```

**Comment by dsyme** [comment]
Fixed in #191
